[Driver] Move some string definitions around (NFC)

modocache opened the pull request:
<!-- Please complete this template before creating the pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->

I've been poking around what `.swiftdeps` files are and how they work, and noticed a small amount of duplication, across `Driver.cpp` and `Compilation.cpp`, in some strings used in the files.

I'm not sure how much of an improvement this is, but these two commits "centralize" the duplication. YAML keys like `version` and `inputs` now exist in `CompilationRecord.h`. Input status identifiers like `!dirty` and `!private` are now defined as part of the `CompileJobAction::InputInfo` struct.

Is it worth making the change? I'm not sure. Feel free to close this pull request if you don't think it is. :)

#### Resolved bug number: None
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

A smoke test on macOS does the following:

1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:

1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

jrose-apple commented on lib/Driver/CompilationRecord.h:
```diff
@@ -0,0 +1,51 @@
+};
+
+/// \returns A string representation of the given key.
+inline StringRef getName(TopLevelKey Key) {
```
Please mark this `static`. Non-static inlines have weird rules that we just avoid.

jrose-apple commented on lib/Driver/CompilationRecord.h:
```diff
@@ -0,0 +1,51 @@
+/// \returns A string representation of the given key.
+inline StringRef getName(TopLevelKey Key) {
+  switch (Key) {
+    case TopLevelKey::Version: return "version";
```
Formatting: LLVM style does not indent `case`.

jrose-apple commented on include/swift/Driver/Action.h:
```diff
@@ -144,6 +146,30 @@ class CompileJobAction : public JobAction {
+
+    /// \returns The string identifier used to represent the given status in a
+    /// compilation record file (.swiftdeps file).
+    static StringRef getStatusName(Status S) {
```
Since this doesn't round-trip cleanly, I think we should give it a different name.

jrose-apple commented on include/swift/Driver/Action.h:
```diff
@@ -144,6 +146,30 @@ class CompileJobAction : public JobAction {
+    /// compilation record file (.swiftdeps file).
+    static StringRef getStatusName(Status S) {
+      switch (S) {
+        case UpToDate:
```
Same formatting comment: no indent for `case`.

jrose-apple commented:
I do feel a little weird putting the build record tags in Action.h, which is theoretically a "public" header. Maybe those belong in CompilationRecord.h as well.

jrose-apple commented:
But eliminating duplication of magic constants is a good idea. The `!private` for .swiftdeps files is probably next, and we can change them all to `!noncascading` at that point.

modocache commented:
@jrose-apple Hopefully I've addressed your comments: I moved everything to `CompilationRecord.h`, did some slight renaming, and fixed up the indentation.

jrose-apple commented on lib/Driver/CompilationRecord.h:
```diff
@@ -0,0 +1,80 @@
+}
+
+/// \returns The string identifier used to represent the given status in a
+/// compilation record file (.swiftdeps file).
```
Last nitpick: can you add a note here pointing out that this will not round-trip?

modocache replied:
Ah, I was actually confused when you mentioned that earlier. By "not round-trip", you mean that `getInfoStatusForIdentifier(getIdentifierForInputInfoStatus(Status)) != Status`, is that right? Since the left-hand side would be an `Optional<Status>`?

modocache replied:
I added a note to the function below, based on my interpretation of "round trip" 😇

jrose-apple replied:
Ah, no, I mostly meant that there was no dedicated identifier for NewlyAdded, and that in general not every input status will necessarily have a unique compilation record identifier.

modocache replied:
Aha, gotcha. My bad! Updating now.

modocache replied:
Done! Let me know if that's what you had in mind.

jrose-apple replied:
Works for me. Thanks, Brian!

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test Linux platform

jrose-apple commented:
@swift-ci Please smoke test Linux platform

jrose-apple commented:
@swift-ci Please smoke test Linux platform

modocache commented:
Looks like all the tests passed, and according to https://github.com/apple/swift/pull/4585#issuecomment-244198577 my merge permissions have been restored. So I'll go ahead and merge this myself. :)
